File: results/0.75/HTTP-1.1/http-1.1-meta-llama-llama-3.3-70b-instruct-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define HAMMERInsensitive 1

void hammer_token(int a, char *b, char *c) {}

int main(int argc, char **argv) {
    if (argc != 2) {
        printf("Usage: %s <file>\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        printf("Failed to open file: %s\n", argv[1]);
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    char *data = malloc(file_size);
    if (!data) {
        printf("Failed to allocate memory\n");
        fclose(file);
        return 1;
    }

    size_t read_size = fread(data, 1, file_size, file);
    if (read_size != file_size) {
        printf("Failed to read file\n");
        free(data);
        fclose(file);
        return 1;
    }

    fclose(file);

    char *id = "[a-zA-Z][a-zA-Z0-9-]*";
    char *token = "[!#$%&'*+.^_`|~0-9a-zA-Z-]+";
    char *separators = "[ ()<>@,;:\\\"/\\[\\]\\?={}\t ]+";
    char *comment = "#.*\n";
    char *quoted_string = "\"[^\"]*\"";
    char *request_line = "(GET|HEAD|POST|PUT|DELETE|OPTIONS|TRACE|CONNECT|PATCH) .*";
    char *request_target = "^[^ ]+|\\*";
    char *http_version = "HTTP/1\\.1";
    char *request = strcat(strdup(request_line), "\\r\\n");
    char *field_name = id;
    char *field_value = "[^\r\n]*";
    char *message_header = strcat(strdup(field_name), " : ");
    message_header = strcat(message_header, field_value);
    message_header = strcat(message_header, "\\r\\n");
    char *headers = strcat(strdup(message_header), "*");
    char *status_code = "^[0-9]{3}";
    char *reason_phrase = "[^\r\n]*";
    char *status_line = strcat(strdup(http_version), " ");
    status_line = strcat(status_line, status_code);
    status_line = strcat(status_line, " ");
    status_line = strcat(status_line, reason_phrase);
    status_line = strcat(status_line, "\\r\\n");
    char *response = strcat(strdup(status_line), headers);
    char *entity_body = ".*";
    char *http_message = strcat(strdup(request), " ");
    http_message = strcat(http_message, headers);
    http_message = strcat(http_message, " ");
    http_message = strcat(http_message, entity_body);
    http_message = strcat(http_message, " | ");
    http_message = strcat(http_message, response);
    http_message = strcat(http_message, " ");
    http_message = strcat(http_message, entity_body);

    printf("%s\n", id);
    printf("%s\n", token);
    printf("%s\n", separators);
    printf("%s\n", comment);
    printf("%s\n", quoted_string);
    printf("%s\n", request_line);
    printf("%s\n", request_target);
    printf("%s\n", http_version);
    printf("%s\n", request);
    printf("%s\n", field_name);
    printf("%s\n", field_value);
    printf("%s\n", message_header);
    printf("%s\n", headers);
    printf("%s\n", status_code);
    printf("%s\n", reason_phrase);
    printf("%s\n", status_line);
    printf("%s\n", response);
    printf("%s\n", entity_body);
    printf("%s\n", http_message);

    free(data);

    return 0;
}

File: results/0.75/HTTP-1.1/http-1.1-meta-llama-llama-3.3-70b-instruct-turbo.rs
use nom::{
    IResult,
    bytes::complete::{tag, take_while1, tag_no_case},
    character::complete::{digit1, multispace0},
    combinator::{map_res},
    multi::{many0},
    sequence::{preceded, tuple},
};
use std::fs::File;
use std::io::Read;
use std::env;
use std::str;

const CRLF: &str = "\r\n";

#[derive(Debug, PartialEq)]
enum HttpMethod {
    OPTIONS,
    GET,
    HEAD,
    POST,
    PUT,
    DELETE,
    TRACE,
    CONNECT,
}

impl HttpMethod {
    fn parse(input: &str) -> IResult<&str, HttpMethod> {
        let (input, method) = tag_no_case("OPTIONS")(input)
            .or_else(|_: nom::Err<nom::error::Error<&str>>| tag_no_case("GET")(input))
            .or_else(|_: nom::Err<nom::error::Error<&str>>| tag_no_case("HEAD")(input))
            .or_else(|_: nom::Err<nom::error::Error<&str>>| tag_no_case("POST")(input))
            .or_else(|_: nom::Err<nom::error::Error<&str>>| tag_no_case("PUT")(input))
            .or_else(|_: nom::Err<nom::error::Error<&str>>| tag_no_case("DELETE")(input))
            .or_else(|_: nom::Err<nom::error::Error<&str>>| tag_no_case("TRACE")(input))
            .or_else(|_: nom::Err<nom::error::Error<&str>>| tag_no_case("CONNECT")(input))?;
        Ok((input, match method {
            "OPTIONS" => HttpMethod::OPTIONS,
            "GET" => HttpMethod::GET,
            "HEAD" => HttpMethod::HEAD,
            "POST" => HttpMethod::POST,
            "PUT" => HttpMethod::PUT,
            "DELETE" => HttpMethod::DELETE,
            "TRACE" => HttpMethod::TRACE,
            "CONNECT" => HttpMethod::CONNECT,
            _ => unreachable!(),
        }))
    }
}

fn http_version(input: &str) -> IResult<&str, &str> {
    tag_no_case("HTTP/1.1")(input)
}

fn request_uri(input: &str) -> IResult<&str, &str> {
    take_while1(|c: char| c != ' ' && c != '\r' && c != '\n')(input)
}

fn request_line(input: &str) -> IResult<&str, (HttpMethod, &str, &str)> {
    tuple((HttpMethod::parse, request_uri, http_version))(input)
}

fn header_name(input: &str) -> IResult<&str, &str> {
    take_while1(|c: char| c.is_alphanumeric() || c == '-' || c == '_')(input)
}

fn header_value(input: &str) -> IResult<&str, &str> {
    take_while1(|c: char| c != '\r' && c != '\n')(input)
}

fn header(input: &str) -> IResult<&str, (&str, &str)> {
    preceded(multispace0, tuple((header_name, tag(": "), header_value)))(input)
}

fn headers(input: &str) -> IResult<&str, Vec<(&str, &str)>> {
    many0(header)(input)
}

fn request(input: &str) -> IResult<&str, (HttpMethod, &str, &str, Vec<(&str, &str)>)> {
    let (input, (method, uri, version)) = request_line(input)?;
    let (input, headers) = headers(input)?;
    Ok((input, (method, uri, version, headers)))
}

fn status_code(input: &str) -> IResult<&str, u16> {
    map_res(digit1, |s: &str| s.parse::<u16>())(input)
}

fn reason_phrase(input: &str) -> IResult<&str, &str> {
    take_while1(|c: char| c != '\r' && c != '\n')(input)
}

fn status_line(input: &str) -> IResult<&str, (u16, &str)> {
    tuple((status_code, reason_phrase))(input)
}

fn response(input: &str) -> IResult<&str, (u16, &str, Vec<(&str, &str)>)> {
    let (input, (code, phrase)) = status_line(input)?;
    let (input, headers) = headers(input)?;
    Ok((input, (code, phrase, headers)))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Usage: {} <input_file>", args[0]);
    }

    let mut file = File::open(&args[1]).unwrap();
    let mut input = Vec::new();
    file.read_to_end(&mut input).unwrap();
    let input_str = str::from_utf8(&input).unwrap();

    let (_remaining, request) = request(input_str).unwrap();
    let (_remaining, response) = response(input_str).unwrap();

    println!("{:?}", request);
    println!("{:?}", response);
}